neuralmagic/sparseml
# [Forkless SparseML Transformers] [Feature Branch] Setting Up The `modification` module

## Summary

This pull request addresses the removal of the dependency on `nm-transformers` in favor of the original HF `transformers.` The primary motivation for this change is to simplify maintenance by eliminating the challenges associated with keeping a fork.

The scope of this PR was defined by running a `diff` between `nm-transformers` and `transformers` and identifying the elements that need to be included in `sparseml` to account for the differences.

For more details refer to the appropriate documentation in the internal docs repository.

## Changes made:
- adding a dependency on `"transformers<4.37"`, removing the dependency on our fork
- creating `src/sparseml/transformers/sparsify/modification` submodule, that is responsible for modifying the `transformers` models, as the fork did in the past
- moving the appropriate scripts from `sparseml.transformers.utils` to `sparseml.transformers.sparsification`
- updating the source code to support the new transformers version (e.g. accounting for renaming `SiLUActivation` to `SiLU`
- adding exhaustive, new unit tests
- fixed the issue of failing `transformers` test in GHA. Refactored and optimized the tests.

## Manual Testing

Tested both `Mistral` and `LlaMa` on our end-to-end flows:
1. Sparsification (one shot + sparse fine-tuning + quantization)
2. Loading the sparsified model and inference
3. Export of the sparsified model

Tested a single BERT QA recipe:
1. Run the sparse fine-tuning
2. Export the model

Note manual testing asserts that the pathways are working. The testing did not check the correctness of the resulting models.

## Relevant PRs:
- [x] #2057
- [x] #2047
- [x] #2050 
- [x] #2051 
- [x] #2052 
- [x] #2054 
- [x] #2056
- [x] #2088
